### PR TITLE
Expose DOTNET_JitPrintInlinedMethods in Release

### PR DESCRIPTION
**Pull request by EgorBo**
Prints a call graph with inliner's decisions for a given method/class/everything (same syntax as for `DOTNET_JitDisasm`)

e.g. `$env:DOTNET_JitPrintInlinedMethods="System.SpanHelpers:*"`: for a hello world 
```
Inlines into System.SpanHelpers:IndexOfNullCharacter(ulong):int:
  [INLINED: aggressive inline attribute] System.SpanHelpers:UnalignedCountVector128(ulong):long
  [INLINED: aggressive inline attribute] System.SpanHelpers:GetCharVector256SpanLength(long,long):long
  [INLINED: aggressive inline attribute] System.SpanHelpers:GetCharVector128SpanLength(long,long):long
  [FAILED: does not return] System.SpanHelpers:ThrowMustBeNullTerminatedString()


Inlines into System.SpanHelpers:IndexOfNullByte(ulong):int:
  [INLINED: aggressive inline attribute] System.SpanHelpers:UnalignedCountVector128(ulong):ulong
  [INLINED: aggressive inline attribute] System.SpanHelpers:GetByteVector256SpanLength(ulong,int):ulong
  [INLINED: aggressive inline attribute] System.SpanHelpers:GetByteVector128SpanLength(ulong,int):ulong
  [FAILED: does not return] System.SpanHelpers:ThrowMustBeNullTerminatedString()


Inlines into System.SpanHelpers:LastIndexOfValueType[short,System.SpanHelpers+DontNegate`1[short]](byref,short,int):int:
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.Int16:System.Numerics.IEqualityOperators<System.Int16,System.Int16,System.Boolean>.op_Equality(short,short):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(bool):bool
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(System.Runtime.Intrinsics.Vector256`1[short]):System.Runtime.Intrinsics.Vector256`1[short]
  [INLINED: aggressive inline attribute] System.SpanHelpers:ComputeLastIndex[short](long,System.Runtime.Intrinsics.Vector256`1[short]):int
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(System.Runtime.Intrinsics.Vector256`1[short]):System.Runtime.Intrinsics.Vector256`1[short]
  [INLINED: aggressive inline attribute] System.SpanHelpers:ComputeLastIndex[short](long,System.Runtime.Intrinsics.Vector256`1[short]):int
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(System.Runtime.Intrinsics.Vector128`1[short]):System.Runtime.Intrinsics.Vector128`1[short]
  [INLINED: aggressive inline attribute] System.SpanHelpers:ComputeLastIndex[short](long,System.Runtime.Intrinsics.Vector128`1[short]):int
  [INLINED: below ALWAYS_INLINE size] System.SpanHelpers+DontNegate`1[short]:NegateIfNeeded(System.Runtime.Intrinsics.Vector128`1[short]):System.Runtime.Intrinsics.Vector128`1[short]
  [INLINED: aggressive inline attribute] System.SpanHelpers:ComputeLastIndex[short](long,System.Runtime.Intrinsics.Vector128`1[short]):int


Inlines into System.SpanHelpers:SequenceCompareTo(byref,int,byref,int):int:
  [INLINED: below ALWAYS_INLINE size] System.Byte:CompareTo(ubyte):int:this
  [INLINED: below ALWAYS_INLINE size] System.Byte:CompareTo(ubyte):int:this
  [INLINED: aggressive inline attribute] System.SpanHelpers:LoadNUInt(byref,ulong):ulong
    [INLINED: aggressive inline attribute] System.Runtime.CompilerServices.Unsafe:ReadUnaligned[ulong](byref):ulong
  [INLINED: aggressive inline attribute] System.SpanHelpers:LoadNUInt(byref,ulong):ulong
    [INLINED: aggressive inline attribute] System.Runtime.CompilerServices.Unsafe:ReadUnaligned[ulong](byref):ulong
  [INLINED: below ALWAYS_INLINE size] System.Byte:CompareTo(ubyte):int:this
System.Byte[]
```

First of all it's useful to catch cases where inliner gives up due to the "budget" thing or callgraph depth, etc so it can quickly highlight problematic cases.
Another bonus that it highlights devirtualized + inlined calls (with PGO)

Since we're making it public we can rename it now if needed.

cc @stephentoub

### CODE REVIEW

File: src/coreclr/jit/fginline.cpp
Comment (AndyAyersMS) [inline src/coreclr/jit/fginline.cpp:1]: There is also code in this method that tries to show failure reasons for calls that were never marked as inline candidates -- do you want to enable/uncomment that too?

**Review by AndyAyersMS — COMMENTED** [review]
I worry that in the more complex cases it will be very hard to correlate this tree back to the source code. The "details" version shows the IL offset which at least gives some help.

Another option we might consider is to annotate the assembly listing with this information, so there's a bit more context.

File: src/coreclr/jit/fginline.cpp
Comment (EgorBo) [reply src/coreclr/jit/fginline.cpp:1]: @AndyAyersMS I'm not sure I understand what code you mean, can you point me at it please?

**Comment by EgorBo** [comment]
> I worry that in the more complex cases it will be very hard to correlate this tree back to the source code. The "details" version shows the IL offset which at least gives some help.
> 
> Another option we might consider is to annotate the assembly listing with this information, so there's a bit more context.

I've just added IL offset. The primary use case is to quickly identify "out of budget" cases because those quite often lead to perf problems because we stop inlining even simple things in hot blocks if exceeded it. It's exactly what happened with Stephen's PR yesterday so this property could help him to identify the problem easily. E.g.:

image

**Comment by EgorBo** [comment]
will address this week

**Comment by EgorBo** [comment]
Ping myself to work on this next week.

**Pull request closed without merging**